Inconsistent tsc behavior on TS 5.4

### 🔎 Search Terms

"tsc inconsistent behavior", "tsc 0 errors", "ts 5.4 bug", "5.4.0-dev.20240219"

### 🕗 Version & Regression Information

I wanted to update TS in our project from `v5.3.3` to `v5.4.3`. And I noticed several weird things:
- When I was switching TS version in VS Code, it wasn't showing any errors. But `tsc --noEmit` was showing 2 errors. I thought that VS Code was just not ready to support `v5.4.3` yet. But the new version comes with `v5.4.3` and it still doesn't work correctly.
- I run `tsc --noEmit`. It shows `2 errors`.  If I open a problematic file and put `// @ts-expect-error`, it shows only `1 error`. But if I remove that line, it instantly shows `0 errors`.

This changed between versions `5.4.0-dev.20240218` and `5.4.0-dev.20240219`.

`5.4.0-dev.20240219` is the first version where I see this problem. I checked `@next` version which is `5.5.0-dev.20240412` and the error is still there.

Unfortunately, I wasn't able to make a reproducible example due complexity of the code. But both errors are about `unknown` type where it should be detected just fine.

### ⏯ Playground Link

_No response_

### 💻 Code

_No response_

### 🙁 Actual behavior

- It doesn't show errors in VS Code
- It shows errors in a console. But after applying `// @ts-expect-error` and removing that line, it shows `0 errors`.

### 🙂 Expected behavior

- Errors in VS Code are detected
- It shows errors in a console. After applying `// @ts-expect-error` and removing that line, it shows all previous errors.

### Additional information about the issue

_No response_

Ok, I was able to reproduce it. Here is a [link](https://stackblitz.com/edit/vitejs-vite-dy48pa?file=src%2Ffail%2Ffail.tsx) to StackBlitz. There are 2 main cases:

1. If you'll open `fail.tsx`, you won't see any errors in the editor. But if you will run `yarn tsc --noEmit` in the terminal, you'll see `src/fail/fail.tsx:18:22 - error TS2322: Type 'unknown[]' is not assignable to type 'FailedItem[]'`.
2. Start `tsc` in watch mode: `yarn tsc --noEmit --watch`. Open `fail.tsx` and add `// @ts-expect-error` in any place of the code. Remove this comment. It will show `0 errors` in the console.

Also you could notice that there is `import '@total-typescript/ts-reset/is-array';` in `reset.d.ts`. The only thing it's doing is changing `any` to `unknown` for `Array.isArray`. [Source code](https://github.com/total-typescript/ts-reset/blob/main/src/entrypoints/is-array.d.ts).

React version doesn't really matter. It works the same on `v16` and `v18`.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -14565,6 +14565,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getApparentTypeOfIntersectionType(type: IntersectionType, thisArgument: Type) {
+        if (type === thisArgument) {
+            return type.resolvedApparentType || (type.resolvedApparentType = getTypeWithThisArgument(type, thisArgument, /*needApparentType*/ true));
+        }
         const key = `I${getTypeId(type)},${getTypeId(thisArgument)}`;
         return getCachedType(key) ?? setCachedType(key, getTypeWithThisArgument(type, thisArgument, /*needApparentType*/ true));
     }

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -14565,7 +14565,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getApparentTypeOfIntersectionType(type: IntersectionType, thisArgument: Type) {
-        return type.resolvedApparentType || (type.resolvedApparentType = getTypeWithThisArgument(type, thisArgument, /*needApparentType*/ true));
+        const key = `I${getTypeId(type)},${getTypeId(thisArgument)}`;
+        return getCachedType(key) ?? setCachedType(key, getTypeWithThisArgument(type, thisArgument, /*needApparentType*/ true));
     }
 
     function getResolvedTypeParameterDefault(typeParameter: TypeParameter): Type | undefined {

--- a/tests/baselines/reference/intersectionApparentTypeCaching.symbols
+++ b/tests/baselines/reference/intersectionApparentTypeCaching.symbols
@@ -1,0 +1,19 @@
+//// [tests/cases/compiler/intersectionApparentTypeCaching.ts] ////
+
+=== intersectionApparentTypeCaching.ts ===
+// https://github.com/microsoft/TypeScript/issues/58175
+
+type TX<T extends any[] & object> = T["length"];
+>TX : Symbol(TX, Decl(intersectionApparentTypeCaching.ts, 0, 0))
+>T : Symbol(T, Decl(intersectionApparentTypeCaching.ts, 2, 8))
+>T : Symbol(T, Decl(intersectionApparentTypeCaching.ts, 2, 8))
+
+type T0<U extends any[] & object> = U;
+>T0 : Symbol(T0, Decl(intersectionApparentTypeCaching.ts, 2, 48))
+>U : Symbol(U, Decl(intersectionApparentTypeCaching.ts, 3, 8))
+>U : Symbol(U, Decl(intersectionApparentTypeCaching.ts, 3, 8))
+
+type T1 = T0<string[]>;
+>T1 : Symbol(T1, Decl(intersectionApparentTypeCaching.ts, 3, 38))
+>T0 : Symbol(T0, Decl(intersectionApparentTypeCaching.ts, 2, 48))
+

--- a/tests/baselines/reference/intersectionApparentTypeCaching.types
+++ b/tests/baselines/reference/intersectionApparentTypeCaching.types
@@ -1,0 +1,17 @@
+//// [tests/cases/compiler/intersectionApparentTypeCaching.ts] ////
+
+=== intersectionApparentTypeCaching.ts ===
+// https://github.com/microsoft/TypeScript/issues/58175
+
+type TX<T extends any[] & object> = T["length"];
+>TX : TX<T>
+>   : ^^^^^
+
+type T0<U extends any[] & object> = U;
+>T0 : U
+>   : ^
+
+type T1 = T0<string[]>;
+>T1 : string[]
+>   : ^^^^^^^^
+

--- a/tests/cases/compiler/intersectionApparentTypeCaching.ts
+++ b/tests/cases/compiler/intersectionApparentTypeCaching.ts
@@ -1,0 +1,8 @@
+// @strict: true
+// @noEmit: true
+
+// https://github.com/microsoft/TypeScript/issues/58175
+
+type TX<T extends any[] & object> = T["length"];
+type T0<U extends any[] & object> = U;
+type T1 = T0<string[]>;

